aten/src/ATen/TensorOptions.h

```diff
@@ -0,0 +1,186 @@
+#pragma once
+
+#include <ATen/Context.h>
+#include <ATen/Device.h>
+#include <ATen/Layout.h>
+#include <ATen/ScalarType.h>
+#include <ATen/Tensor.h>
+#include <ATen/Type.h>
+#include <ATen/optional.h>
+
+#include <cstddef>
+
+namespace at {
+
+/// A class to encapsulate construction axes of a `Tensor`.
+/// `TensorOptions` is a virtual class to enable overriding of certain methods
+/// by subclasses in other libraries, such as PyTorch. In PyTorch, there is a
+/// `torch::TensorOptions` subclass of this `TensorOptions`, which changes
+/// `type()` to return a variable type instead of a tensor type, such that
+/// variables are created inside factory methods, instead of tensors.
+struct TensorOptions {
+  /// Constructs the `TensorOptions` with valid defaults, which are:
+  /// - dtype: float
+  /// - device: CPU
+  /// - layout: strided
+  TensorOptions() = default;
+
+  /// Constructs the `TensorOptions` from the type of the given `Tensor`.
+  /// If the `Tensor` has a CUDA type, the `device_index` will match that of the
+  /// tensor. See the constructor from `Type` for the semantics w.r.t. the
+  /// `type()` method.
+  explicit TensorOptions(Tensor tensor, bool discard_runtime_type = false);
+
+  /// Constructs the `TensorOptions` from a type and optional `device_index`.
+  ///
+  /// If `discard_runtime_type` is false (the default), the behavior of
```

Comment:
I'm okay with this, but I wonder if there's a cleaner way.  You really need the type only to know if it's a variable (and because even if you knew this you can't recover it?).  But whether it's a variable or not is already part of the public ATen API.  We can't take something that allows us to do this while still retaining the API we "want" -- layout, device, requires_grad (why is requires_grad not here)?

Reply:
Would `std::optional<bool> requires_grad` work? If it is not present, it is a Tensor, otherwise it is a variable with the equivalent requires_grad setting.

Reply:
Sounds reasonable but @goldsborough mentioned elsewhere he had trouble with optional and NVCC.